File: app/src/androidTest/java/de/westnordost/streetcomplete/quests/AddMaxSpeedTest.java
package de.westnordost.streetcomplete.quests;

import de.westnordost.streetcomplete.data.osm.OsmElementQuestType;
import de.westnordost.streetcomplete.data.osm.changes.StringMapEntryAdd;
import de.westnordost.streetcomplete.data.osm.changes.StringMapEntryModify;
import de.westnordost.streetcomplete.quests.max_speed.AddMaxSpeed;
import de.westnordost.streetcomplete.quests.max_speed.AddMaxSpeedForm;

public class AddMaxSpeedTest extends AOsmElementQuestTypeTest
{
	private static final String MAXSPEED_TYPE = "maxspeed:type";

	public void testNoSign()
	{
		bundle.putString(AddMaxSpeedForm.MAX_SPEED_IMPLICIT_ROADTYPE, "flubberway");
		bundle.putString(AddMaxSpeedForm.MAX_SPEED_IMPLICIT_COUNTRY, "XX");
		verify(new StringMapEntryAdd(MAXSPEED_TYPE,"XX:flubberway"));
	}

	public void testNormalSign()
	{
		bundle.putString(AddMaxSpeedForm.MAX_SPEED, "123");
		verify(
				new StringMapEntryAdd("maxspeed","123"),
				new StringMapEntryAdd(MAXSPEED_TYPE,"sign"));
	}

	public void testAdvisoryNormalSign()
	{
		bundle.putString(AddMaxSpeedForm.ADVISORY_SPEED, "123");
		verify(
				new StringMapEntryAdd("maxspeed:advisory","123"),
				new StringMapEntryAdd(MAXSPEED_TYPE+":advisory","sign"));
	}

	public void testZoneSign()
	{
		bundle.putString(AddMaxSpeedForm.MAX_SPEED, "123");
		bundle.putString(AddMaxSpeedForm.MAX_SPEED_IMPLICIT_ROADTYPE, "zoneXYZ");
		bundle.putString(AddMaxSpeedForm.MAX_SPEED_IMPLICIT_COUNTRY, "AA");
		verify(
				new StringMapEntryAdd("maxspeed","123"),
				new StringMapEntryAdd(MAXSPEED_TYPE,"AA:zoneXYZ"));
	}

	public void testLivingStreet()
	{
		tags.put("highway","residential");
		bundle.putBoolean(AddMaxSpeedForm.LIVING_STREET, true);
		verify(
				new StringMapEntryModify("highway","residential","living_street"));
	}

	@Override protected OsmElementQuestType createQuestType()
	{
		return new AddMaxSpeed(null);
	}
}

File: app/src/androidTest/java/de/westnordost/streetcomplete/quests/AddMaxSpeedTest.kt
package de.westnordost.streetcomplete.quests

import de.westnordost.streetcomplete.data.osm.changes.StringMapEntryAdd
import de.westnordost.streetcomplete.data.osm.changes.StringMapEntryModify
import de.westnordost.streetcomplete.data.osm.download.OverpassMapDataDao
import de.westnordost.streetcomplete.quests.max_speed.AddMaxSpeed
import de.westnordost.streetcomplete.quests.max_speed.AddMaxSpeedForm
import org.junit.Test

import org.mockito.Mockito.mock

class AddMaxSpeedTest : AOsmElementQuestTypeTest() {

    override val questType = AddMaxSpeed(mock(OverpassMapDataDao::class.java))

    @Test fun noSign() {
        bundle.putString(AddMaxSpeedForm.MAX_SPEED_IMPLICIT_ROADTYPE, "flubberway")
        bundle.putString(AddMaxSpeedForm.MAX_SPEED_IMPLICIT_COUNTRY, "XX")
        verify(StringMapEntryAdd(MAXSPEED_TYPE, "XX:flubberway"))
    }

    @Test fun normalSign() {
        bundle.putString(AddMaxSpeedForm.MAX_SPEED, "123")
        verify(
            StringMapEntryAdd("maxspeed", "123"),
            StringMapEntryAdd(MAXSPEED_TYPE, "sign")
        )
    }

    @Test fun advisoryNormalSign() {
        bundle.putString(AddMaxSpeedForm.ADVISORY_SPEED, "123")
        verify(
            StringMapEntryAdd("maxspeed:advisory", "123"),
            StringMapEntryAdd("$MAXSPEED_TYPE:advisory", "sign")
        )
    }

    @Test fun zoneSign() {
        bundle.putString(AddMaxSpeedForm.MAX_SPEED, "123")
        bundle.putString(AddMaxSpeedForm.MAX_SPEED_IMPLICIT_ROADTYPE, "zoneXYZ")
        bundle.putString(AddMaxSpeedForm.MAX_SPEED_IMPLICIT_COUNTRY, "AA")
        verify(
            StringMapEntryAdd("maxspeed", "123"),
            StringMapEntryAdd(MAXSPEED_TYPE, "AA:zoneXYZ")
        )
    }

    @Test fun livingStreet() {
        tags["highway"] = "residential"
        bundle.putBoolean(AddMaxSpeedForm.LIVING_STREET, true)
        verify(StringMapEntryModify("highway", "residential", "living_street"))
    }

    companion object {
        private val MAXSPEED_TYPE = "maxspeed:type"
    }
}
